envoy-control-tests/src/main/resources/lua_spec/ingress_handler_spec.lua

```diff
@@ -0,0 +1,227 @@
+              "\nUnauthorized request: method = , path = , clientIp = , clientName = , protocol = https, statusCode = ")
+            assert.spy(handle.logInfo).was_called(1)
+        end)
+
+        it("request with no metadata saved", function ()
+            -- given
+            metadata['envoy.filters.http.lua'] = nil
+            headers = {}
+            local handle = handlerMock(headers, metadata, ssl)
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_called_with(_,
+              "\nUnauthorized request: method = , path = , clientIp = , clientName = , protocol = https, statusCode = ")
+            assert.spy(handle.logInfo).was_called(1)
+        end)
+
+        it("request with empty path", function ()
+            -- given
+            metadata['envoy.filters.http.lua']['request.info.path'] = ''
+            local handle = handlerMock(headers, metadata, ssl)
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_called_with(_,
+              "\nUnauthorized request: method = POST, path = '', "..
+                "clientIp = 127.1.1.3, clientName = service-first, protocol = https, statusCode = 403")
+            assert.spy(handle.logInfo).was_called(1)
+        end)
+    end)
+
+    describe("should not log requests:", function()
+
+        it("request with no rbac metadata", function()
+            -- given
+            metadata = {}
+            local handle = handlerMock(headers, metadata, ssl)
+            local metadataMock = handle:streamInfo():dynamicMetadata()
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_not_called()
+            assert.spy(metadataMock.get).was_not_called_with(_, 'envoy.filters.http.lua')
+        end)
+
+        it("authorized request", function()
+            -- given
+            metadata['envoy.filters.http.rbac']['shadow_engine_result'] = 'allowed'
+            local handle = handlerMock(headers, metadata, ssl)
+            local metadataMock = handle:streamInfo():dynamicMetadata()
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_not_called()
+            assert.spy(metadataMock.get).was_not_called_with(_, 'envoy.filters.http.lua')
+        end)
+    end)
+
+    it("should handle x-forwarded-for formats", function ()
+        local xff_to_expected_client_ip= {
+            {"", ""},
+            {"127.9.3.2", "127.9.3.2"},
+            {"3.23.2.44 , 2.34.3.2,127.1.3.5", "127.1.3.5"},
+            {"2001:db8:85a3:8d3:1319:8a2e:370:7348,1001:db8:85a3:8d3:1319:8a2e:370:2222", "1001:db8:85a3:8d3:1319:8a2e:370:2222"},
+            {"2001:db8:85a3:8d3:1319:8a2e:370:7348,127.1.3.4", "127.1.3.4"}
+        }
+
+        for i,v in ipairs(xff_to_expected_client_ip) do
```

Comment:
Is there a way to run parametrised tests in Lua? This way if one of the cases fail the whole test fails so we have to run a debugger to figure things out.

Reply:
I didn't find any way to make parametrised tests in Busted: https://olivinelabs.com/busted/ and this is most popular and up-to-date lua test framework (but it still lacks a documentation a lot!).

> This way if one of the cases fail the whole test fails
We can change the assertion to not fail the whole `it(...)` test, but I don't know if this will be useful. Unfortunately `was_called_with` assertion shows only this message when assertion fails with message:
> Function was not called with the arguments
not showing the actual arguments

BTW: "Easy to setup" and "working as expected" debugger don't exists for Lua as far as I know :)

Reply:
Since those tests are just code couldn't we move this loop outside ` it("should handle x-forwarded-for formats"` and generate tests like that?

Reply:
It's great idea, implemented. Now we can see exactly which format has failed. Thanks.